### PR TITLE
Add Linux Flatpak build

### PR DESCRIPTION
**Pull request by ntninja**
Ported the engine to Linux Flatpak based on some of the AppImage work. I tried to make sure that Flatpak and AppImage share as much data (AppStream, .desktop, build scripts) as possible. Last commit even shares list of archives to download and verify between the two, but it is entirely optional.

Snapcraft should probably be using the common data as well, but I just didn't feel like touching that thing. (Just some personal dismay towards that ecosystem from my side. :stuck_out_tongue:)

### CODE REVIEW

**Comment by ntninja** [comment]
Last commit fixes display of window icon on Wayland (where `SDL_SetWindowIcon` doesn't work). This is much more visible with the Flatpak as SDL will always use Wayland there if it is available due to the `--socket=fallback-x11` sandbox permission denying X11 access in those cases.

**Comment by isage** [comment]
Can you change appid to org.nxengine.nxengine_evo?
Other than that - looks good.

**Comment by ntninja** [comment]
@isage: It is done, but dang I loved having the name `sarcasticat` part of the App-ID! :slightly_smiling_face:

**Comment by ntninja** [comment]
(Tiny update: FreeDesktop Runtime/SDK version 20.08 was release a couple of days ago, so the Flatpak now uses that instead 19.08.)

**Comment by timsueberkrueb** [comment]
> but I just didn't feel like touching that thing

Haha, that's fine. I will look into it. I, on the other hand, have no idea about Flatpak ;)

**Comment by isage** [comment]
> @isage: It is done, but dang I loved having the name `sarcasticat` part of the App-ID!

Yeah, but `sarcasticat` is my alt-nickname and nxevo is build with code from many contributors, so...)


That aside, it would be nice to have Flatpak/AppImage automatic builds on appveyour (same as with others, build and upload to releases on tag)

**Comment by ntninja** [comment]
@timsueberkrueb: Great! Be sure to ping me if there are any change you think could impact the Flatpak or AppImage builds.

As a quick overview over my reusable changes:

  * `platform/xdg/` contains the XDG metadata (AppStream metadata, .desktop file and icon) with the agreed upon App-ID (`org.nxengine.nxengine_evo`). As these files reference each other by that App-ID, they cannot be renamed.
  * `build-scripts/unix.install-common.sh` contains a scripts that installs all the data files and XDG metadata into the given DEST path (the only command-line parameter). On Flatpak this path is `/app`, on a direct host install it would be `/usr` or `/usr/local`. I'm pretty sure it should be useful for Snappy as well  – just read the top comment in that script as it is a bit picky about its environment to keep it minimal.

Cheers! :slightly_smiling_face:

**Comment by ntninja** [comment]
@isage: I figured that be the case, but still… :grin: 

> That aside, it would be nice to have Flatpak/AppImage automatic builds on appveyour (same as with others, build and upload to releases on tag)

Is that a feature-request you have towards me? :eyes:

**Comment by isage** [comment]
Yep

**Comment by timsueberkrueb** [comment]
@ntninja I looked into it. Snapcraft, the meta build tool for snaps, supports AppStream metadata. However, since we are providing nightly snap builds using the [public snapcraft build infrastructure](https://snapcraft.io/build), the AppStream metdadata would need to be generated dynamically. Currently, snapcraft extracts the version information from git (see [snapcraft.yaml](https://github.com/nxengine/nxengine-evo/blob/master/.snapcraft.yaml#L57)). Do you think it would be reasonable to create a script that automatically creates an AppStream file for this purpose? I assume this might also be helpful to automate Flatpak and AppImage builds using the CI.

**Comment by timsueberkrueb** [comment]
Btw, @isage, the snap package still lives under my personal name in the snap store (https://snapcraft.io/nxengine-evo). You could create a new store account (e.g. called `nxengine`) to emphasize that the snap package is official. We can transfer the package to the new account and you could add me (or anyone interested in maintaining the snap package) as a collaborator to manage releases.

**Comment by ntninja** [comment]
@timsueberkrueb: I'm not entirely sure why you want it, but *as I understand it* you're asking me if I know some way to automatically add a new release to the AppStream file?

If so, I just did all the work on this for you. :slightly_smiling_face: 

Below is a small XSLT file that can automatically add a new release to an AppStream file given the new version and optionally a date. You'd use it like this (drop the second *stringparam* if you don't need/want the date):

```sh
xsltproc --stringparam "VERSION" "your_version" --stringparam "DATE" "$(date +%Y-%m-%d)" appdata-add-version.xslt org.nxengine.nxengine_evo.appdata.xml
```

```xslt
<?xml version="1.0" encoding="UTF-8"?>
<xsl:stylesheet version="1.0" xmlns:xsl="http://www.w3.org/1999/XSL/Transform">
	<xsl:output method="xml"/>
	<xsl:param name="DATE"/>
	
	<xsl:template match="/component/releases">
		<xsl:copy>
			<release>
				<xsl:attribute name="version">
					<xsl:value-of select="$VERSION"/>
				</xsl:attribute>

				<xsl:if test="boolean($DATE)">
					<xsl:attribute name="date">
						<xsl:value-of select="$DATE"/>
					</xsl:attribute>
				</xsl:if>
			</release>
			<xsl:apply-templates select="@* | node()" />
		</xsl:copy>
	</xsl:template>

	<xsl:template match="@* | node()">
		<xsl:copy>
			<xsl:apply-templates select="@* | node()" />
		</xsl:copy>
	</xsl:template>
</xsl:stylesheet>
```

**Comment by timsueberkrueb** [comment]
@ntninja cool, thank you :)

> I'm not entirely sure why you want it

I was considering whether it would be possible to reuse the AppStream file for building the snap. Snapcraft can extract information from AppStream files. However, it extracts everything, including the version number. As we are providing nightly snap builds, I would need some way to change this version number at build-time to indicate that the version being built is a nightly build rather than a release. I was assuming that this might be relevant for nightly AppImage or Flatpak builds as well although I may be wrong (?) .

I can also try to find another way to override the version number and we could also not reuse the AppStream file for snapcraft if it doesn't provide much of an advantage.

What do you think?

**Comment by ntninja** [comment]
> I can also try to find another way to override the version number and we could also not reuse the AppStream file for snapcraft if it doesn't provide much of an advantage.
>
> What do you think?

There is nothing inherently wrong with this way of adding a release to the AppStream data. I'd even argue it's one of cleanest ways to do so you'll ever find (if I didn't already know some XSLT, I may had used RegExp instead). If it is useful to you I'd definitely recommend doing something similar to what I suggested there. :stuck_out_tongue: 

Either way, I very much think that we should strive to deduplicate as much of the build process between the different Linux/Unix distribution methods as possible. Particularily for the AppStream file with its long non-technical strings, somebody might come along and fix a typo in one file and then they'd become out of sync…

TL;DR: Unless you cannot generate the modifed AppStream file before SnapCraft processes it, I'd recommend reusing it just for the sake of deduplication if nothing else.

> I was assuming that this might be relevant for nightly AppImage or Flatpak builds as well although I may be wrong (?) .

I wasn't really planning on providing beta/nightly builds actually, but for releases this way of bumping versions would be something I'll likely use, yes.

**Comment by timsueberkrueb** [comment]
> Either way, I very much think that we should strive to deduplicate as much of the build process between the different Linux/Unix distribution methods as possible. Particularily for the AppStream file with its long non-technical strings, somebody might come along and fix a typo in one file and then they'd become out of sync…

Yeah, that's what I was thinking as well. Will see what I (or, snapcraft, rather) can do :)
